File: c++/merge/merge.cpp
#include <iostream>
using namespace std;

int tmp[10010] = {};

void mergeSort(int nums[], int left, int right) {
    if (left >= right) {
        return;
    }
    int mid = (left + right) >> 1;
    mergeSort(nums, left, mid);
    mergeSort(nums, mid + 1, right);
    int i = left, j = mid + 1, k = 0;
    while (i <= mid && j <= right) {
        if (nums[i] <= nums[j]) {
            tmp[k++] = nums[i++];
        } else {
            tmp[k++] = nums[j++];
        }
    }
    while (i <= mid) {
        tmp[k++] = nums[i++];
    }
    while (j <= right) {
        tmp[k++] = nums[j++];
    }
    for (i = left, j = 0; i <= right; ++i, ++j) {
        nums[i] = tmp[j];
    }
}

int main() {
    int n;
    cin>>n;
    int nums[n] = {};
    for (int i = 0; i < n; ++i) {
        cin>>nums[i];
    }
    mergeSort(nums, 0, n - 1);
    for (int i = 0; i < n; ++i) {
        printf("%d ", nums[i]);
    }
    system("pause");
    return 0;
}

File: python/merge.py
def mergeSort(nums, left, right):
    if left >= right:
        return
    mid = (left + right) >> 1
    mergeSort(nums, left, mid)
    mergeSort(nums, mid+1, right)
    tmp = []
    i, j = left, mid + 1
    while i <= mid and j <= right:
        if nums[i] <= nums[j]:
            tmp.append(nums[i])
            i += 1
        else:
            tmp.append(nums[j])
            j += 1
    while i <= mid:
        tmp.append(nums[i])
        i += 1
    while j <= right:
        tmp.append(nums[j])
        j += 1
    j = 0
    for i in range(left, right + 1):
        nums[i] = tmp[j]
        j += 1


if __name__ == '__main__':
    n = int(input())
    nums = list(map(int, input().split()))
    mergeSort(nums, 0, n-1)
    print(' '.join(list(map(str, nums))))
